lib/Sema/CSRanking.cpp

```diff
@@ -962,6 +965,32 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
       }
     }
 
+    // Swift 4.1 compatibility hack: If everything else is considered equal,
+    // favour a property on a concrete type over a protocol property member.
+    //
+    // This hack is required due to changes in shadowing behaviour where a
+    // protocol property member will no longer shadow a property on a concrete
+    // type, which created unintentional ambiguities in 4.2. This hack ensures
+    // we at least keep these cases unambiguous in Swift 5 under Swift 4
+    // compatibility mode. Don't however apply this hack for decls found through
+    // dynamic lookup, as we want the user to have to disambiguate those.
+    //
+    // This is intentionally narrow in order to best preserve source
+    // compatibility under Swift 4 mode by ensuring we don't introduce any new
+    // ambiguities. This will become a more general "is more specialised" rule
+    // in Swift 5 mode.
+    if (!tc.Context.isSwiftVersionAtLeast(5) &&
+        choice1.getKind() != OverloadChoiceKind::DeclViaDynamic &&
+        choice2.getKind() != OverloadChoiceKind::DeclViaDynamic &&
+        isa<VarDecl>(decl1) && isa<VarDecl>(decl2)) {
+      auto *nominal1 = dc1->getSelfNominalTypeDecl();
+      auto *nominal2 = dc2->getSelfNominalTypeDecl();
+      if (nominal1 && nominal2 && nominal1 != nominal2) {
+        isSwift4ConcreteOverProtocolVar1 = isa<ProtocolDecl>(nominal2);
+        isSwift4ConcreteOverProtocolVar2 = isa<ProtocolDecl>(nominal1);
```

Comment:
It seems like these two lines would be a little clearer if they were actually looking at the decl associated with each variable rather than the other one, e.g.:

```
isSwift4ConcreteOverProtocolVar1 = !isa<ProtocolDecl>(nominal1);
isSwift4ConcreteOverProtocolVar2 = !isa<ProtocolDecl>(nominal2);
```

Reply:
And perhaps rename the LHS of each. I mention this because at first I thought I was looking at typos in the name on the RHS.

Reply:
@rudkx That's actually how I originally wrote it, but wasn't sure the extra visual noise was worth it – I'll swap it back :). Any suggestions for the naming of the LHS? I can't immediately think of a better alternative.

Reply:
@hamishknight Perhaps drop `isVarAndNotProtocol{1,2}`?

I don't think keeping `Swift4` in the name adds a lot of meaning and it's pretty clear that it's only set under `-swift-version 4` so I don't think there's going to be a lot of confusion over what it's for.

Reply:
@rudkx I agree that's better, thanks! I've gone ahead and made that change.